src/modifiers/sub_token.py

```diff
@@ -7,32 +7,47 @@
 tag: user.cursorless
 """
 
+mod.list("cursorless_subtoken", desc="Supported subcomponent types")
+ctx.lists["self.cursorless_subtoken"] = {"word": "word", "char": "character"}
 
-@mod.capture(rule=("<user.ordinals> | last"))
+
+@mod.capture(rule="<user.ordinals> | last")
 def ordinal_or_last(m) -> str:
     """An ordinal or the word 'last'"""
     if m[0] == "last":
         return -1
     return m.ordinals - 1
 
 
-mod.list("cursorless_subtoken", desc="Supported subcomponent types")
+@mod.capture(rule="<user.ordinal_or_last> [past <user.ordinal_or_last>]")
+def cursorless_ordinal_range(m) -> str:
+    """Ordinal range"""
+    return {
+        "startIndex": m.ordinal_or_last_list[0],
+        "endIndex": m.ordinal_or_last_list[-1] + 1 or None,
+    }
 
-ctx.lists["self.cursorless_subtoken"] = {"word": "word", "char": "character"}
+
+@mod.capture(rule="(first | last) <number_small>")
+def cursorless_first_last_range(m) -> str:
+    """First/last range"""
+    if m[0] == "first":
+        return {"startIndex": 0, "endIndex": m.number_small}
+    return {"startIndex": -m.number_small, "endIndex": -1}
```

Comment:
I believe `endIndex` should be `None` here like with `last` above. Then I don't think you'll need the PR extension-side

Reply:
We do because -2 and the like need to work aswell. Also reversed range. I did extensive testing and the changes are needed as far as I can tell.

Reply:
That's not a bug; endIndex is not inclusive, and semantics of negative endIndex is that it gets added to length. Make sense?

Reply:
Ok so I went back and looked at your other PR and the surrounding code, and I agree that something is off.  I'll take a quick swing at it

Reply:
I agree with you. I just got into my head that I wanted to be able to do a complete slice using negative end index. But that not how it works in python for example.

I have now updated both pr using null for end index and all seems to work fine.

Reply:
Yeah so what if we had it so that start is always less than end, and we just pass `isReversed: true` if necessary?

Reply:
In the latest PR end is already exclusive an all works fine as far as I can test. I think its better to leave that logic to the extension and the talon side just gives us the ranges as it is.

Reply:
So in latest PR it looks like we have `endIndex+1`. If `endIndex<startIndex`, I believe we technically should have `endIndex-1` to make it exclusive, no? But that would be really confusing to me, so I think it would be more easily understandable to have the range always be forward and use `isReversed`

Reply:
The +1 is gone. Now its end exclusive for negative ranges as well. We do endIndex -1 already so problem solved I would say.

https://github.com/pokey/cursorless-vscode/blob/bd5dd5317919e274cfe79efa184e7472adee1b07/src/processTargets.ts#L244

Reply:
Sorry [this](https://github.com/pokey/cursorless-talon/blob/1b4e1a83138ece41d3f9c7ab106e69d5fc883e40/src/modifiers/sub_token.py#L27) is the `+1` I'm talking about.  If I said "take third past first subword", we'd get `{startIndex: 3, endIndex: 2}`, whereas in my mind exclusive would be `{startIndex: 3, endIndex: 0}`.  So I think it's just simpler to say `{startIndex: 0, endIndex: 4, isReversed: true}`, which is the same semantics that VSCode has with `Selection`: the `start` is always before `end`, but it is the `anchor` and `active` that are swapped.

Alternately, we could just make the ranges inclusive on both sides and leave the start / end index reversed.  That might make our lives easier, even tho it goes against conventions

Idk @brxck wanna weigh in here?  I'm not in love with any of these options tbh but I don't see anything better 😅

Reply:
`{startIndex: 2, endIndex: 1}` is end exclusive already since that will give us endIndex 0 extension side. Doing -1 as endIndex wont work ofcourse. 

Everything already works so my point is that we don't actually have to change anything if we don't want to and keep its as simple as possible talon side is always nice I think. We can do isReversded but it will give us a bit more logic talon side and I don't really see the gain.

I vote for just leaving it as it is because it already works as expected.

Reply:
> `{startIndex: 2, endIndex: 1}` is end exclusive already since that will give us endIndex 0 extension side.

I'm not sure I follow this... Maybe that's an indicator of sorts.

In the case of a reversed selection, if `endIndex = 1` gives us the first word, that doesn't seem exclusive? I would expect to get the second word if inclusive, and the second if exclusive. This seems like a confusing behavior, am I thinking about this wrong? 

I don't have a strong preference between your solutions pokey, but it's hard to fault mirroring the VS Code API with something like `isReversed`

Reply:
Its exclusive but also always 0 indexed. That the end index is smaller than start index have no affect on the api until its evaluated in vscode but Talon doesn't actually have to care about that if we don't want to. If you guys prefere isReversed then please go ahead with it. For now I just want any changes to be merged so we can do reversed selection regardless of implementation.

Reply:
Ok I do see the logic in your version now @AndreasArvidsson, ie that the fact that start is greater than end has no impact on the code

That being said, it does break the definition of "exclusive". "Exclusive" means that the given index is not an element of the range. But in your version, if `endIndex` is 1, and `endIndex<startIndex`, then `endIndex` **will** be included in the range, so by definition `endIndex` is not exclusive

I do agree it would be good to get this stuff merged, but the api contract between talon and extension is important to get right, as that's where we'll be defining our tests and is a natural boundary.

I'm happy to wire up the `isReversed` version, but given I'm on holiday I'm not sure exactly when that will be

Reply:
Yes it only exclusive in regards to the end index which is not a correct definition, but as I said reverse range slices is not really something you often see.

I can do it, but i think i still would prefer inclusive index instead because the makes the logic easier talon side, but it doesn't really matter.

Reply:
So when in doubt I usually go with principle of least surprise. Ie if I see the structure, will I be surprised when I find out what it actually does. Advantage of `isReversed` is that I don't think it's going to be surprising when you find out what it does. The inclusive thing isn't too bad, tho I do think people might be surprised when for the forward case the end index is inclusive, as I think it's a pretty common convention for endIndex to be exclusive

That being said, I do like the symmetry and simplicity of the inclusive version. @brxck Tiebreak? Still throwing your weight behind `isReversed`?

Reply:
I completely agree with that guiding sentiment. Probably would anchorIndex and activeIndex help with that since peoples expectations of start and end ranges wouldn't apply. I think that's my 5000cent. The decision is on @brxck :D

Reply:
The pressure 😂 if it reduces complexity, then my vote would be for the symmetry of making everything inclusive. I feel like a simpler implementation is worth more than following a relatively benign convention here.

Reply:
And the Emmy goes to...(drum roll)...inclusive indices!

@AndreasArvidsson I leave it up to you whether to switch to anchor / active or stick with start / end. I could go either way

Fwiw inclusive is actually consistent with how we handle normal ranges, so I think that's a good sign. 

I guess there we use start / end, so if we change to anchor / active here, we should prob switch that there as well. If so, we can leave that for the target overhaul work I'm starting on soon

So we're all on the same page, I believe we've aligned on "take third past first word" maps to `{anchor: 2, active: 0}` and "take first past third word" maps to `{anchor: 0, active: 2}`

Reply:
Agreed. I'll look into that tomorrow and update existing PRs :)

Reply:
@pokey @brxck 
And changes are done! :)